src/sql/workbench/contrib/objectExplorer/browser/serverTreeView.ts

```diff
@@ -232,66 +298,83 @@ export class ServerTreeView extends Disposable implements IServerTreeView {
+	 * @param connectionId The connection ID to search for
+	 */
+	private getConnectionInTreeInput(connectionId: string): ConnectionProfile | undefined {
+		if (this._tree instanceof AsyncServerTree) {
+			const root = this._tree.getInput();
+			const connections = ConnectionProfileGroup.getConnectionsInGroup(root);
+			return connections.find(conn => conn.id === connectionId);
+		} else {
+			const root = TreeUpdateUtils.getTreeInput(this._connectionManagementService);
+			const connections = ConnectionProfileGroup.getConnectionsInGroup(root);
+			const results = connections.filter(con => {
+				if (connectionId === con.id) {
+					return true;
+				} else {
+					return false;
+				}
+			});
+			if (results && results.length > 0) {
+				return results[0];
 			}
-		});
-		if (results && results.length > 0) {
-			return results[0];
+			return undefined;
 		}
-		return null;
 	}
 
-	private onObjectExplorerSessionCreated(connection: IConnectionProfile) {
-		const conn = this.getConnectionInTreeInput(connection.id);
-		if (conn) {
-			this._tree.refresh(conn).then(() => {
-				return this._tree.expand(conn).then(() => {
-					return this._tree.reveal(conn, 0.5).then(() => {
-						this._treeSelectionHandler.onTreeActionStateChange(false);
-					});
-				});
-			}).then(null, errors.onUnexpectedError);
+	private async onObjectExplorerSessionCreated(connection: IConnectionProfile): Promise<void> {
+		const element = this.getConnectionInTreeInput(connection.id);
+		if (element) {
+			if (this._tree instanceof AsyncServerTree) {
+				this._tree.rerender(element);
+			} else {
+				await this._tree.refresh(element);
+			}
+			await this._tree.expand(element);
+			await this._tree.reveal(element, 0.5);
+			this._treeSelectionHandler.onTreeActionStateChange(false);
 		}
 	}
 
 	public addObjectExplorerNodeAndRefreshTree(connection: IConnectionProfile): void {
 		hide(this.messages);
 		if (!this._objectExplorerService.getObjectExplorerNode(connection)) {
-			this._objectExplorerService.updateObjectExplorerNodes(connection).then(() => {
-				// The oe request is sent. an event will be raised when the session is created
-			}, error => {
-			});
+			this._objectExplorerService.updateObjectExplorerNodes(connection).catch(e => errors.onUnexpectedError(e));
 		}
 	}
 
-	public deleteObjectExplorerNodeAndRefreshTree(connection: IConnectionProfile): Promise<void> {
+	public async deleteObjectExplorerNodeAndRefreshTree(connection: IConnectionProfile): Promise<void> {
 		if (connection) {
 			const conn = this.getConnectionInTreeInput(connection.id);
 			if (conn) {
-				return this._objectExplorerService.deleteObjectExplorerNode(conn).then(async () => {
+				await this._objectExplorerService.deleteObjectExplorerNode(conn);
+				if (this._tree instanceof AsyncServerTree) {
+					// Collapse the node before refreshing so the refresh doesn't try to fetch
+					// the children again (which causes it to try and connect)
+					this._tree.collapse(conn);
```

Comment:
what's the difference between the asynchronous collapse in the else statement and the synchronous one here? And why does it need to be synchronous?

Reply:
Two different types of trees. The AsyncServerTree has a synchronous collapse function while the old ITree version is async. (you'll see these kind of small differences all over - just some of the many things that the two trees do differently)